Support for egressgateway IPv6 policies in the controlplane

rgo3 opened the pull request:
This PR is a follow up to https://github.com/cilium/cilium/pull/37713, where we introduced datapath handling for IPv6 egress gateway policies. The goal of this PR is to make sure the agent creates the correct bpf state for the datapath, in terms of bpf map entries in the IPv6 egress gateway map. Please see individual commits for more implementation details.

To provide an example, a EgressGatewayPolicy can now look like the following:
```
apiVersion: cilium.io/v2
kind: CiliumEgressGatewayPolicy
metadata:
  name: egress-sample
spec:
  # Specify which pods should be subject to the current policy.
  # Multiple pod selectors can be specified.
  selectors:
  - podSelector:
      matchLabels:
        org: empire
        class: mediabot
        # The following label selects default namespace
        io.kubernetes.pod.namespace: default
  # Specify which destination CIDR(s) this policy applies to.
  # Multiple CIDRs can be specified.
  destinationCIDRs:
  - "0.0.0.0/0"
  - "::/0"

  # Configure the gateway node.
  egressGateway:
    # Specify which node should act as gateway for this policy.
    nodeSelector:
      matchLabels:
        node.kubernetes.io/name: node2
    # The first IPv4 or IPv6 address assigned to this interface, will be chosen as the egress IP respectively
    interface: "eth0"
```

This policy would create entries in both the IPv4 and IPv6 policy maps, given it specifies destination CIDRs for both protocols.

```release-note
Add support for IPv6 egress gateway policies
```

rgo3 committed:
egressgateway: refactor egressmap interface

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 committed:
pkg/egressgateway: add ipv6 euqivalent bpf map userspace wrapper

This commit adds the necessary golang datastructures and API to access
the ipv6 egressgateway policy bpf map from the cilium agent code.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 commented on pkg/egressgateway/policy.go:
```diff
@@ -278,6 +290,23 @@ func ParseCEGP(cegp *v2.CiliumEgressGatewayPolicy) (*PolicyConfig, error) {
+			return nil, fmt.Errorf("failed to parse egress IP %s: %w", egressGateway.EgressIP, err)
+		}
+
+		// Check if egressIP address family matches all destination and excluded CIDRs
```
If an egress IP is specified, all other IPs specified in the CRD must match the same protocol version (v4 or v6).

rgo3 replied:
Removed support for v6 egressIP for now.

rgo3 commented on pkg/maps/egressmap/policy.go:
```diff
@@ -77,31 +92,42 @@ func createPolicyMapFromDaemonConfig(in struct {
 	}
 
-	out.MapOut = bpf.NewMapOut(PolicyMap(createPolicyMap(in.Lifecycle, in.PolicyConfig, ebpf.PinByName)))
+	//if in.EnableIPv6EgressGateway {
```
Still need to add a new feature flag. Should this translate into the datapath? Currently the datapath only uses one feature Macro to compile the datapath code.

julianwiedmann replied:
You mean that if IPv6 is disabled for the agent, then we also don't need to do the lookup in the datapath? I think we can leave that as future optimization - along with "if IPv4 is disabled, don't do an IPv6 lookup".

rgo3 commented on pkg/egressgateway/policy.go:
```diff
@@ -157,35 +160,52 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
 
 	switch {
 	case gc.iface != "":
```
This is where I run into the first "hickups" with my current approach. Technically, to support v6 policies, the CRD doesn't need to change much. This also means that with one EgressGatewayPolicy, both v4 and v6 policies can be created, given a PodSelector selects both v4 and v6 workloads and the selected interface can have v4 and v6 addresses assigned. However, this makes reporting errors a bit more complicated, e.g. if the environment is v4 only, we shouldn't throw an error because we couldn't find an assigned v6 address and vice versa respectively. 

@msune suggested to now only throw an error if we found no IP at all, but this would make it harder to indicate real failure if we were only able to retrieve one v4 or v6 address in a dual stack environment.

rgo3 commented on pkg/egressgateway/manager.go:
```diff
@@ -577,46 +591,94 @@ func (manager *Manager) relaxRPFilter() error {
+		logger := log.WithFields(logrus.Fields{
+			logfields.SourceIP:        endpointIP,
+			logfields.DestinationCIDR: dstCIDR.String(),
+			logfields.EgressIP:        gwc.egressIP4,
```
Also log v6 egress ip.

rgo3 reviewed (commented):
This is a first draft of this PR and there are still some things missing but currently it is a bit unclear to me how the controlplane should decide on when to create or retrieve and assign version dependent resources if an egress gateway policy crd object is ip version agnostic. And what the failure mode should be if a policy is technically valid for both ip stacks but the controlplane was only able to partially apply it. Any feedback would be appreciated!

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -51,6 +53,14 @@ var (
+	// IPv6 special values
+	// ExcludedCIDRIPv6 is a special IP value used as gatewayIP in the BPF policy map
+	// to indicate the entry is for an excluded CIDR and should skip egress gateway
+	ExcludedCIDRIPv6 = netip.MustParseAddr("::1")
```
Given that `gatewayIP` is still an IPv4 address (we're not doing IPv6 underlay yet), I suppose we're not using this part yet?

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -577,46 +591,94 @@ func (manager *Manager) relaxRPFilter() error {
+	if manager.policyMap4 != nil {
+		manager.policyMap4.IterateWithCallback(
+			func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
+				egressPolicies4[*key] = *val
```
Could we do this as separate `updateEgressRulesV4()` and `updateEgressRulesV6()` ?

I appreciate it means we process every endpoint twice, but it would be much easier to review. And we can fine-tune later.

julianwiedmann reviewed (commented):
Thank you! A few comments inline. I think once we have `--enable-ipv6-egress-gateway` wired up and disable it in the e2e CLI tests, we could merge this fairly soon. And then connect the last dots, so that all variants of policies etc work.

rgo3 committed:
pkg/egressgateway: support ipv6 addresses in gateway configs

This commit allows users to define IPv6 addresses as destination and
excluded CIDRs in EgressGatewayPolicies, which internally will cause
IPv6 egress IPs to be derived for the internal representation of a
gateway.

Setting a IPv6 egress IP directly is not supported yet.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 committed:
pkg/egressgateway: reconcile ipv6 gateway configs into bpf state

This commit adds reconciliation for IPv6 policies into bpf state in the
datapath. Meaning for IPv6 workloads and specified destination and
exclusion CIDRs the controlplane code inserts the correct IPv6 egress
gateway policy entries into the bpf map. It should be noted that, to
redirect traffic to a gateway node, cilium still used a IPv4 underlay,
meaning the gateway node IP of a IPv6 egress gateway policy entry is
still IPv4.

This commit also extends the unit test suite for the egress gateway
manager to test the handling of IPv6 policies in the same conditions as
we test IPv4 egress gateway policies.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -566,6 +571,7 @@ func (manager *Manager) relaxRPFilter() error {
 		ifaceName := pc.gatewayConfig.ifaceName
 		if _, ok := ifSet[ifaceName]; !ok {
 			ifSet[ifaceName] = struct{}{}
+			// Does this also need to happen for ipv6?
```
I'd expect the `ipv4` key to be a global control, and https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt doesn't list an `ipv6` variant.

rgo3 replied:
Alright, removed the comment then.

julianwiedmann commented on pkg/egressgateway/policy.go:
```diff
@@ -268,6 +302,9 @@ func ParseCEGP(cegp *v2.CiliumEgressGatewayPolicy) (*PolicyConfig, error) {
 			return nil, fmt.Errorf("failed to parse destination CIDR %s: %w", cidrString, err)
 		}
 		dstCidrList = append(dstCidrList, cidr)
+		if cidr.Addr().Is6() {
+			policyGwc.v6needed = true
+		}
```
Ah that's a nice trick! This is then something we could in the future also do with IPv4, for an IPv6-only installation ?

rgo3 replied:
Yes, I think so. Just didn't do it because it currently does not apply.

julianwiedmann commented on pkg/k8s/apis/cilium.io/v2/cegp_types.go:
```diff
@@ -50,15 +52,15 @@ type CiliumEgressGatewayPolicySpec struct {
 	// Should be a subset of destinationCIDRs otherwise it will not have any
 	// effect.
 	//
 	// +kubebuilder:validation:Optional
-	ExcludedCIDRs []IPv4CIDR `json:"excludedCIDRs"`
+	ExcludedCIDRs []CIDR `json:"excludedCIDRs"`
```
I bet we also need to bump `CustomResourceDefinitionSchemaVersion` for this, so that the new CRD version is picked up on upgrade.

rgo3 replied:
Just do double check, but I'll have to bump https://github.com/cilium/cilium/blame/83398a81351ebee5f0bf3b8105cdee4ae7c3eca0/pkg/k8s/apis/cilium.io/register.go#L18 to 1.31.8 correct?

julianwiedmann replied:
yep!

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -589,8 +604,18 @@ func (manager *Manager) updateEgressRules() {
+			logfields.DestinationCIDR: dstCIDR.String(),
+			logfields.EgressIP:        gwc.egressIP4,
+			logfields.GatewayIP:       gwc.gatewayIP,
+		})
```
Did you intentionally move this block up? Just checking as we would previously potentially log the `GatewayIP` as `ExcludedCIDRIPv4` below. But now we always log `gwc.gatewayIP`.

rgo3 replied:
Unintentional, fixed.

julianwiedmann reviewed (commented):
A few more comments inline from browsing through, nothing controversial.

pippolo84 commented on cilium-dbg/cmd/bpf_egress_list.go:
```diff
@@ -38,7 +38,7 @@ var bpfEgressListCmd = &cobra.Command{
 		common.RequireRootPrivilege("cilium bpf egress list")
 
-		policyMap, err := egressmap.OpenPinnedPolicyMap()
+		policyMap, err := egressmap.OpenPinnedPolicyMap4()
```
Shouldn't we extend `cilium-dbg` to list the entries from PolicyMap6 too?

rgo3 replied:
Have a follow-up PR (#38691) that adds this as I need it for connectivity tests.

pippolo84 commented on pkg/datapath/linux/netdevice/netdevice.go:
```diff
@@ -71,3 +71,62 @@ func getIfaceWithIPv4Address(ip netip.Addr) (string, error) {
+
+func GetIfaceWithIPv6Address(ip netip.Addr) (string, error) {
+	return getIfaceWithIPv6Address(ip)
+}
```
Nit: no need to define an inner unexported function `getIfaceWithIPv6Address`. I see that we are doing it for IPv4 but that's because the inner function is called by `TestForIfaceWithIPv4Address` too.

rgo3 replied:
Will fix that up, but on another note it looks like `TestForIfaceWithIPv4Address` is unused, so maybe we can do the same refactoring for the ipv4 funcs as a follow up.

pippolo84 commented on pkg/egressgateway/policy.go:
```diff
@@ -157,37 +161,65 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
 		// If the gateway config specifies an egress IP, use the interface with that IP as egress
 		// interface
-		gwc.egressIP = gc.egressIP
+		gwc.egressIP4 = gc.egressIP
```
Nit: might want to leave a comment here to state the specifying an IPv6 is not supported yet

pippolo84 commented on pkg/egressgateway/policy.go:
```diff
@@ -157,37 +161,65 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
-			gwc.egressIP = EgressIPNotFoundIPv4
-			return fmt.Errorf("failed to find interface with default route: %w", err)
+			gwc.egressIP4 = EgressIPNotFoundIPv4
+			return fmt.Errorf("failed to find interface with IPv4 default route: %w", err)
```
Shouldn't we try to go ahead with the IPv6 codepath in case of an IPv4 error? If so, we could call both paths independently of the respective results and then combine the errors with `errors.Join`.

rgo3 replied:
Currently looking at this code path and it doesn't seem super beneficial to allow further execution if previous calls are failing, as it's very likely the following calls will fail as well within the ipv4 path but also for the ipv6 path as we want to retrieve the same device, and ipv6 only environment won't be supported for egress gateway yet. For that reason I would find using errors.Join just makes the code slightly less readable.

pippolo84 reviewed (commented):
Nice! :tada: 

Overall LGTM, left some questions and a couple of nits inline.

rgo3 commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -653,7 +747,7 @@ func waitForReconciliationRun(tb testing.TB, egressGatewayManager *Manager, curr
 		}
 
-		time.Sleep(10 * time.Millisecond)
+		time.Sleep(30 * time.Millisecond) // worried that I need to change this, what's the huge performance blocker here?
```
Remembered this when addressing some feedback, and want to highlight it since it can easily get lost. I'm still a bit confused why I had to significantly raise the timeout. Anyone has a theory as to why? cc @julianwiedmann

julianwiedmann replied:
I've seen this as well in the past, when experimenting with map migration for the EGW policy map (effectively managing two policy maps in parallel). We now reconcile two maps, so the total reconciliation takes longer. I don't have a good feeling whether the extra time is expected / excessive ...

rgo3 replied:
Made a TODO comment to investigate why the timeout increase is necessary.

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -197,6 +201,10 @@ func NewEgressGatewayManager(p Params) (out struct {
+	if dcfg.EnableIPv6 && !dcfg.EnableIPv6Masquerade {
+		return out, fmt.Errorf("egress gateway requires --%s=\"true\" if --%s=\"true\"", option.EnableIPv6Masquerade, option.EnableIPv6Name)
+	}
+
```
I could see this additional requirement causing unexpected issues for users on upgrade (they haven't changed anything in their config, but suddenly the agent refuses to start).

Let's take a follow-up item to make this more graceful (caching the lack of sufficient IPv6 support during startup, and later rejecting any CEGP that lists IPv6 DstCIDRs?). Along with some controlplane testing that demonstrates the desired behavior for an IPv4-only cluster with IPv6 DstCIDRs.

rgo3 replied:
I actually removed the checks and made a TODO note for future improvements, will create follow-up issues for the remaining TODO items I marked in the code.

julianwiedmann reviewed (approved):
lgtm! One follow-up TODO, but not blocking.

pippolo84 reviewed (approved):
:rocket:
